src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -257,16 +288,18 @@ void addPartitionSensors(int partition) {
    */
   private class ProduceRunnable implements Runnable {
     private final int _partition;
+    private final String _key;
 
-    ProduceRunnable(int partition) {
+    ProduceRunnable(int partition, String key) {
       _partition = partition;
+      _key = key;
     }
 
     public void run() {
       try {
-        long nextIndex = _nextIndexPerPartition.get(_partition).get();
+        long nextIndex = _nextIndexPerKey.get(_key).get();
         String message = Utils.jsonFromFields(_topic, nextIndex, System.currentTimeMillis(), _producerId, _recordSize);
-        BaseProducerRecord record = new BaseProducerRecord(_topic, _partition, null, message);
+        BaseProducerRecord record = new BaseProducerRecord(_topic, null, _key, message);
```

Comment:
Here is another reason to include both partition and key. If we include only the key, we have to make sure that producer always use the same partitioning algorithm that is specified in the config and used by `ProduceService`. It is necessary but inconvenient to do so. For example, user have to change this algorithm to switch producer class from old producer and new producer. Things go wrong if they forgot to do so. We can avoid this inconvenience if user only wants to monitor one cluster. We can not avoid this inconvenience for user that wants to monitor a pipeline, and that inconvenience may be justified since user has to keep the producer's partition algorithm in-sync with MM's partitioning algorithm anyway.